src/useContainerSize/useContainerSize.ts

```diff
@@ -0,0 +1,171 @@
+import {
+  useCallback,
+  useEffect,
+  useLayoutEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react';
+import { throttle } from '@utils';
+
+export type Breakpoints = Record<string, number>;
+
+export interface UseContainerSizeOptions<T> {
+  /**
+   * Breakpoints map where key is a breakpoint name
+   * and value is a minimum container width in pixels.
+   */
+  breakpoints: T;
+  /**
+   * Flag that enables or disables hook logic.
+   * When disabled the hook does not listen to size changes.
+   */
+  isEnabled?: boolean;
+  /**
+   * Throttle time in milliseconds for resize events.
+   */
+  throttleTime?: number;
+}
+
+export interface UseContainerSizeResult<T extends Breakpoints> {
+  /**
+   * Current breakpoint name or null if no breakpoint matches.
+   */
+  size: keyof T | null;
+  /**
+   * Current container width in pixels.
+   */
+  width: number | null;
+  /**
+   * Ref callback for the tracked element.
+   */
+  ref: (node: HTMLElement | null) => void;
+}
+
+/**
+ * React hook that tracks container width and returns a breakpoint name
+ * based on the provided breakpoints map.
+ * The hook uses ResizeObserver under the hood and throttles resize events.
+ */
+
+export const useContainerSize = <T extends Breakpoints>(
+  options: UseContainerSizeOptions<T>
+): UseContainerSizeResult<T> => {
+  const { breakpoints, isEnabled = true, throttleTime = 10 } = options;
+
+  const [element, setElement] = useState<HTMLElement | null>(null);
+
+  const parsedBreakpoints = useMemo(() => {
+    return Object.entries(breakpoints)
+      .map(([name, value]) => {
+        return { name, value };
+      })
+      .sort((a, b) => {
+        return a.value - b.value;
+      });
+  }, [breakpoints]);
+
+  const getCurrentSize = useCallback(
+    (width: number): keyof T | null => {
```

Comment:
Let's just create a function outside a hook (in the same module) and pass all required variables, so function will be pure.